### PR TITLE
Made pattern-matcher diagnostics lazily reported + added TORCH_COMPILE_CPROFILE

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110504
* #110501

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by mlazos — APPROVED** [review]

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -50,8 +50,13 @@
...
 NodeOrConstant = Union[Constant, torch.fx.Node]
 
+
+class Multiple:
+    pass
+
+
 # Sentinel indicating multiple quantities can be matched
-MULTIPLE = object()
+MULTIPLE = Multiple()
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:59]: why was this needed, just curious ?

**Review by eellison — APPROVED** [review]
looks good!

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (Chillee) [reply torch/_inductor/pattern_matcher.py:59]: Just for typechecking reasons. It woulda caught the prettyprinter error (although I guess I'm not actually using the prettyprinter here haha).

Comment (eellison) [inline torch/_inductor/pattern_matcher.py:59]: it's a little bit confusing that `Multiple` and `MULTIPLE` are now both in scope. 

Can you not use `MULTIPLE` the sentinel object as a type annotation ?

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -308,7 +321,7 @@ def __repr__(self):
         return f"{self.__class__.__name__}({self.fns_repr()})"
 
     def has_multiple_users(self) -> bool:
-        return self.users is MULTIPLE or self.users > 1
+        return isinstance(self.users, Multiple) or self.users > 1
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:324]: This is less clear `isinstance(self.users, Multiple)` than what existed previously `self.users is MULTIPLE`

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (Chillee) [reply torch/_inductor/pattern_matcher.py:59]: No you can't. I also tried something like `Literal[object]`, but was also unable to get it to work.

File: torch/_inductor/pattern_matcher.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/pattern_matcher.py:324]: Unfortunately, I don't see how to get this to typecheck with mypy other than this.

**Comment by PaliC** [comment]
@pytorchbot revert -m "temporarily will revert as it's causing problems with difftrain import" -c ghfirst

Afterwards will rebase, and force merge
